[Relay][QNN] Simulated Quantize and Dequantize

jwfromm opened the pull request:
This PR adds `simulated_quantize` and `simulated_dequantize` to the QNN library in relay. These operators are primarily meant to support the pass-based quantization framework proposed [in this Discuss post](https://discuss.tvm.apache.org/t/rfc-quantization-quantization-in-tvm/9161). However, these new ops can be cleanly broken into their own PR and can be useful for other applications. The obvious benefit of simulated qnn ops is that they mimic real quantization in floating point. The more interesting benefit of this approach is that it allows switching between per-channel and scalar QNN parameters and changing the datatype **without recompilation**. This has major compute time benefits when doing calibration or eventually trying to do quantization aware training.

I also found that using qnn quantize and dequantize with per-channel parameters and negative axes caused an error and fixed it and changed a test case to catch it going forward.

jwfromm commented:
@electriclilies @anijain2305 @masahi @mbrookhart can you guys take a look at this PR?

jwfromm commented on python/tvm/topi/nn/qnn.py:
```diff
@@ -0,0 +1,186 @@
+    "int32": SQNN_INT32,
+}
+
+SQNN_CODE_TO_DTYPE = {v: k for k, v in SQNN_DTYPE_TO_CODE.items()}
```
Note that the use of integer codes to map to datatypes is a hack since relay doesn't currently support string variables. Once it does, this can be simplified.

anijain2305 commented:
@jwfromm Thanks for the contribution. One high-level question is - Is it possible to have one op - say `simulate_q` that has both `in_dtype` and `out_dtype` params and can act as `simulated_quantize` or `simulated_dequantize` depending on dtypes? The input/output will always be of dtype fp32, but the `in_dtype` and `out_dtype` params can be int/float and tell if our intent is to simulate quantize or dequantize.

jwfromm commented:
I think we could do that, yes, but I would argue that having a clear analogue to `qnn.quantize` and `qnn.dequantize` is conceptually much cleaner. It also has some benefits when moving from simulation to real qnn through pattern matching.

anijain2305 commented:
I see, yeah that should be ok. 

Maybe it's not relevant to this PR but there is a slight catch about `requantize`. I was hoping that the `requantize` could also be represented using the `simulate_q` (or `simualted_requantize` op) where both `in_dtype` and `out_dtype` are integers (while the actual in/out tensors are fp32). The issue representing `requantize` with a sequence of `quantize-dequantize` is that requantize's integer-only computation gives different results compared to the sequence of quantize-dequantize (the deviation grows further for <8 bits). If there was just one op, then we can simulate `requantize` with hopefully more fidelity (should help QAT as well in future).

But I totally understand that it can make design complicated and maybe also difficult to pattern match. Please give it a thought and see if this is helpful. Otherwise, please feel free to ignore.

anijain2305 commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -143,6 +178,39 @@ def dequantize(data, input_scale, input_zero_point, axis=-1):
 
 
+def simulated_dequantize(data, input_scale, input_zero_point, axis=-1, in_dtype="int8"):
+    r"""Simulated Quantize op
```
s/Quantize/Dequantize/

anijain2305 commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -143,6 +178,39 @@ def dequantize(data, input_scale, input_zero_point, axis=-1):
 
+def simulated_dequantize(data, input_scale, input_zero_point, axis=-1, in_dtype="int8"):
+    r"""Simulated Quantize op
+    Mimics the quantize op but has more flexibility in valid inputs and always
```
s/quantize/dequantize/ --> couple of more places in the doc string

anijain2305 commented on python/tvm/topi/nn/qnn.py:
```diff
@@ -0,0 +1,186 @@
+        A scalar tensor representing the scale to use when quantizing to integer datatypes.
+        When it contains more than a single value, N must match the number of channels in data.
+
+    output_zero_point: tvm.te.Tensor, optional
```
Typically, the zero points are scalar. Even for asymmetric, they are scalar. This is done mostly for performance reasons. But, since these ops are generic, it's better to keep it the way that you have.

anijain2305 commented on python/tvm/topi/nn/qnn.py:
```diff
@@ -0,0 +1,186 @@
+    # Use an if chain to dynamically return the proper quantization based on the input datatype.
+    # This allows the op to compile once but apply different quantization approaches
+    # using a variable datatype input.
+    def _dispatch_sim_quantize(value):
```
+1, clever trick.

anijain2305 commented on python/tvm/topi/nn/qnn.py:
```diff
@@ -0,0 +1,186 @@
+@tvm.te.tag_scope(tag=topi.tag.ELEMWISE)
+def simulated_dequantize(data, in_dtype, input_scale=None, input_zero_point=None, axis=-1):
+    """Simulated QNN dequantize operator that mimics QNN outputs in floating point. The benefit
+    of this operator over true QNN quantize is that this operator allows dynamic datatype
```
s/quantize/dequantize

anijain2305 commented on src/relay/qnn/op/simulated_quantize.cc:
```diff
@@ -0,0 +1,82 @@
+
+TVM_REGISTER_NODE_TYPE(SimulatedQuantizeAttrs);
+
+bool SimulatedQuantizeRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
```
Minor suggestion (feel free to ignore) - How about sharing the TypeRel functions for Quantize and Dequantize? they seem to be same.

jwfromm replied:
A solid suggestion but to do that cleanly I'd have to put both simulated ops in one file which would break parity with how the regular quantize and dequantize ops are written. I think I slightly prefer a little code duplication but clearer structure.

anijain2305 reviewed (approved):
LGTM, with minor comments

mbrookhart commented on python/tvm/topi/nn/qnn.py:
```diff
@@ -0,0 +1,186 @@
+        The channel axis for quantization. Default value is -1 which corresponds to the last axis.
+
+    """
+    # Since all simulated outputs are in float32, we can just return the input tensor for fp32.
```
I'm not sure I understand this. Shouldn't we still shift and scale with float inputs?

anijain2305 replied:
Basically act as a passthrough in case you dont want to (de)quantize?

jwfromm replied:
Yeah exactly, this is allowing us to turn off quantization / dequantization if we want to.

mbrookhart replied:
I don't think we should be doing this based off dtype. As you mentioned in the type relation function, we might want to pass in something that isn't float32 and run this against it's own dtype. What's wrong with allowing the user to pass in scale=1, zp=0, dtype=data.dtype if they want to passthrough?

mbrookhart commented on python/tvm/topi/nn/qnn.py:
```diff
@@ -0,0 +1,186 @@
+
+    """
+    # Since all simulated inputs are in float32, we can just return the input tensor for fp32.
+    def _compute_fp32(value, *indices):
```
Same as above, shouldn't we still shift and scale?

mbrookhart commented on src/relay/qnn/op/simulated_dequantize.cc:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  // assign output type
+  reporter->Assign(types[4], TensorType(data->shape, data->dtype));
```
If the output of dequantize is fp32, doesn't this assume the input is always fp32? What if I had a senario where I was trying to simulate the quantization of int32->int8, and the dequantization of int8->int32?

jwfromm replied:
maybe i should clarify the docs. There's no need for the inputs outputs to explicitly be float32. The simulated ops will return whatever the input data type is. I think this is a good behavior to have since it lets them be inserted into any graph without introducing type issues.

anijain2305 replied:
@mbrookhart are you talking about requantize operation?

For your example of quantization of int32 -> int8, is the input a quantized tensor with scale and zero point, or just a plain int32 tensor.

* If it is just a plain int32 tensor, should we even quantize it? From definition standpoint, the quantize (dequantize) has input (output) always as `float32` datatype.

* However, if the input is a quantized integer representation, then you are doing a `requantize` operation (which in this case can be represented by a sequence of simulated_quantize - simulated_dequantize op).

mbrookhart replied:
I was thinking just a plain int32 input, not a quantized version. I'm not sure if we'll hit this in real models, but the possibility is always there, and I'd rather not make assumptions about inputs.

mbrookhart commented on src/relay/qnn/op/simulated_quantize.cc:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  // assign output type
+  reporter->Assign(types[4], TensorType(data->shape, data->dtype));
```
Same confusing about input types as above

mbrookhart reviewed (commented):
Some minor concerns about floating value defaults. I might be missing something.

jwfromm commented:
@mbrookhart I cleaned up the docs to make the datatype behavior more clear and changed `SQNN_FP32` to `SQNN_DISABLE` to make it more clear why you would use it. What do you think?

mbrookhart reviewed (approved):
LGTM

jwfromm commented:
@ZihengJiang I'd love to hear your take before this gets merged.

jwfromm commented:
@masahi it would also be great to hear what you think about this implementation of simulated qnn.

masahi commented:
@jwfromm Looks great, thanks!

masahi commented:
thanks @jwfromm @anijain2305 @mbrookhart
